### PR TITLE
Rust crate preparation - final touches

### PR DESCRIPTION
**Pull request by NlightNFotis**
This is the final PR before we publish the crate on `crates.io`.

Have slightly improved docs, build and added a test to verify that our translation
of C++ exceptions to Rust Result types work.

`cargo test`, `cargo build`, `cargo doc` and `cargo publish --dry-run` all work at this point.

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/libcprover-rust/readme.md [hunk 1]
```diff
@@ -27,7 +29,7 @@ directory of the CBMC project.)
 ```sh
 $ cd src/libcprover-rust
 $ cargo clean
-$ CBMC_LIB_DIR=../../build/lib CBMC_VERSION=5.78.0 cargo build
+$ CBMC_INCLUDE_DIR=../../build/include CBMC_LIB_DIR=../../build/lib CBMC_VERSION=5.78.0 cargo build
```
Comment (TGWDB) [inline src/libcprover-rust/readme.md:32]: Maybe don't fix a version of CBMC here, use something like `CBMC_VERSION=X.XX.X`?

File: src/libcprover-rust/readme.md [hunk 2]
```diff
@@ -36,7 +38,118 @@ look like this:
 ```sh
 $ cd src/libcprover-rust
 $ cargo clean
-$ CBMC_LIB_DIR=../../build/lib CBMC_VERSION=5.78.0 cargo test -- --test-threads=1 --nocapture
+$ CBMC_INCLUDE_DIR=../../build/include CBMC_LIB_DIR=../../build/lib CBMC_VERSION=5.78.0 cargo test -- --test-threads=1 --nocapture
```
Comment (TGWDB) [inline src/libcprover-rust/readme.md:41]: And here.

File: src/libcprover-rust/readme.md (hunk 1 above)
Comment (NlightNFotis) [reply src/libcprover-rust/readme.md:32]: This is in the documentation, and I meant it as a concrete example. 

Should I still change this? Perhaps a good middle ground might be to add a reference to the version schema in the description of the environment variable, starting in line 19.

Comment (TGWDB) [reply src/libcprover-rust/readme.md:32]: I'm agnostic on it. On the one hand I can see someone simply copying and pasting, on the other hand a concrete example is nicer than one with further things to modify.

File: src/libcprover-rust/include/c_errors.h [hunk 3]
```diff
@@ -9,10 +9,11 @@
 // in that we don't want to export internal headers to the clients, and our
 // current build system architecture on the C++ end doesn't allow us to do so.
 //
-// At the same time, we want to allow the Rust API to be able to catch at the shim
-// level the errors generated within CBMC, which are C++ types (and subtypes of
-// those), and so because of the mechanism that cxx.rs uses, we need to have the
-// types present at compilation time (an incomplete type won't do - I've tried).
+// At the same time, we want to allow the Rust API to be able to catch at the
+// shimlevel the errors generated within CBMC, which are C++ types (and
```
Comment (TGWDB) [inline src/libcprover-rust/include/c_errors.h:13]: shimlevel -> shim level

File: src/libcprover-rust/src/lib.rs [hunk 4]
```diff
@@ -101,6 +101,22 @@ mod tests {
...
+    // library depends on, and it will be correctly translated into the Result type for Rust.
+    // This also validates that our type definition include of the base class for the exceptions
+    // works as we expect it to.
+    #[test]
+    fn it_translates_exceptions_to_errors() {
+        let client = cprover_api::new_api_session();
+
+        // The vector of string is supposed to contain a string denoting
+        // a filepath that is erroneous.
+        let vec: Vec<String> = vec!["/fkjsdlkjfisudifoj2309".to_owned()];
```
Comment (TGWDB) [inline src/libcprover-rust/src/lib.rs:114]: I have not tried to break this, but what happens if I have a valid input file at `/fkjsdlkjfisudifoj2309`? I'm also curious if the `/` is OS specific to any degree?

**Review by TGWDB — APPROVED** [review]

File: src/libcprover-rust/src/lib.rs (hunk 4 above)
Comment (NlightNFotis) [reply src/libcprover-rust/src/lib.rs:114]: Well, that was my best attempt at making a random string + file path combo, without resorting to a more elaborate mechanism based on random strings or whatnot.

I'm personally happy with it failing if a user does have a *randomly* (?) named file locally present. If that happens, the user/developer can just have a look at the test, and see that we are looking for this file, inspect his filesystem and locate the file, and potentially alter the test/delete the file.

**Pull request merged**